File: Day_02/day02.py
#!/usr/bin/env python3

INPUT = "./input.txt"

def first_part():
    valid_passwords = 0
    with open(INPUT,"r") as f:        
        for line in f:
            freq_table = {}
            l = line.split(':')
            password = l[1]
            rules = l[0].split()
            min = int(rules[0].split('-')[0])
            max = int(rules[0].split('-')[1])
            key_char = rules[1]

            for c in password:
                freq_table[c] = freq_table.get(c,0) + 1
            if freq_table.get(key_char,0) in range(min,max+1):
                valid_passwords += 1

    print(valid_passwords)

def second_part():
    valid_passwords = 0
    with open(INPUT,'r') as f:
        for line in f:        
            l = line.split(':')
            password = l[1]
            rules = l[0].split()
            min = int(rules[0].split('-')[0])
            max = int(rules[0].split('-')[1])
            key_char = rules[1]
            found = False

            for i,c in enumerate(password):            
                if i == min and c == key_char:
                    found = True
                if i == max and c == key_char:
                    if found:
                        found = False
                    else:
                        found = True
            if found:
                valid_passwords += 1
    print(valid_passwords)


def main():
    first_part()
    second_part()

if __name__ == "__main__":
    main()

File: Day_02/day02.java
package Day_02;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class day02 {
    public static void main(String[] args) {
        firstPart();
        secondPart();
    }

    public static void firstPart() {
        var br = new BufferedReader(new InputStreamReader(day02.class.getResourceAsStream("input.txt")));
        String line;
        int valids = 0;
        try {

            while ((line = br.readLine()) != null) {
                var input = line.split(":");
                var rule = input[0];
                var password = input[1];
                var ruleParts = rule.split(" ");
                var min = Integer.parseInt(ruleParts[0].split("-")[0]);
                var max = Integer.parseInt(ruleParts[0].split("-")[1]);
                var ch = ruleParts[1].charAt(0);

                var times = password.codePoints().filter(i -> i == ch).count();

                if (min <= times && times <= max) {
                    valids++;
                }
            }

            System.out.println("A valid jelszavak száma: " + valids);
            br.close();

        } catch (Exception e) {
            // TODO: handle exception
        }

    }

    public static void secondPart() {
        var br = new BufferedReader(new InputStreamReader(day02.class.getResourceAsStream("input.txt")));
        String line;
        int valids = 0;
        try {
            while ((line = br.readLine()) != null) {
                var input = line.split(":");
                var rule = input[0];
                var password = input[1];
                var ruleParts = rule.split(" ");
                var min = Integer.parseInt(ruleParts[0].split("-")[0]);
                var max = Integer.parseInt(ruleParts[0].split("-")[1]);
                var ch = ruleParts[1].charAt(0);
                int found = 0;

                if (password.charAt(min) == ch) {
                    found++;
                }
                if(password.charAt(max) == ch) {
                    found++;
                }

                if(found == 1) {
                    valids++;
                }
            }

            System.out.println("2. A valid jelszavak száma: " + valids);
            br.close();
        } catch (Exception e) {
            // TODO: handle exception
        }
    }
}
